File: rust/src/validation/schema_discovery.rs
// #exonware/xwsystem/rust/src/validation/schema_discovery.rs
//! Schema validator discovery (plugin entry points).
//! 
//! This module provides a lightweight "plugin" mechanism for discovering
//! implementations of `xwsystem.validation.contracts.ISchemaValidator`.
//! 
//! In Rust, we use a registry pattern instead of Python's entry points.

use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};

use super::contracts::ISchemaValidator;

/// Constant: DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP
pub const DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP: &str = "xwsystem.schema_validators";

/// Structured discovery information for observability/debugging.
#[derive(Debug, Clone)]
pub struct SchemaValidatorDiscoveryResult {
    pub group: String,
    pub loaded: HashMap<String, String>,  // name -> provider repr
    pub errors: HashMap<String, String>,  // name -> error string
}

impl SchemaValidatorDiscoveryResult {
    pub fn new(group: String) -> Self {
        Self {
            group,
            loaded: HashMap::new(),
            errors: HashMap::new(),
        }
    }
}

// Global registry for schema validators
static SCHEMA_VALIDATOR_REGISTRY: OnceLock<Mutex<SchemaValidatorRegistry>> = OnceLock::new();

fn get_registry() -> &'static Mutex<SchemaValidatorRegistry> {
    SCHEMA_VALIDATOR_REGISTRY.get_or_init(|| Mutex::new(SchemaValidatorRegistry::new()))
}

struct SchemaValidatorRegistry {
    providers: HashMap<String, HashMap<String, Arc<dyn ISchemaValidator + Send + Sync>>>,
    default_provider: HashMap<String, String>,
    attempted_groups: std::collections::HashSet<String>,
}

impl SchemaValidatorRegistry {
    fn new() -> Self {
        Self {
            providers: HashMap::new(),
            default_provider: HashMap::new(),
            attempted_groups: std::collections::HashSet::new(),
        }
    }
}

/// Discover schema validator providers.
pub fn discover_schema_validators(
    group: Option<&str>,
    force: bool,
) -> (HashMap<String, Arc<dyn ISchemaValidator + Send + Sync>>, SchemaValidatorDiscoveryResult) {
    let group_name = group.unwrap_or(DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP);
    let mut registry = get_registry().lock().unwrap();
    
    let already_attempted = registry.attempted_groups.contains(group_name);
    if already_attempted && !force {
        let providers = registry.providers.get(group_name)
            .cloned()
            .unwrap_or_default();
        let loaded_repr: HashMap<String, String> = providers.iter()
            .map(|(k, v)| (k.clone(), format!("{:?}", v)))
            .collect();
        return (
            providers,
            SchemaValidatorDiscoveryResult {
                group: group_name.to_string(),
                loaded: loaded_repr,
                errors: HashMap::new(),
            },
        );
    }

    // In Rust, we can't dynamically discover entry points like Python
    // Providers must be manually registered via set_schema_validator
    let providers = registry.providers.get(group_name)
        .cloned()
        .unwrap_or_default();
    let loaded_repr: HashMap<String, String> = providers.iter()
        .map(|(k, v)| (k.clone(), format!("{:?}", v)))
        .collect();

    registry.attempted_groups.insert(group_name.to_string());
    
    // Preserve existing default selection if still valid; else choose first
    if !registry.default_provider.contains_key(group_name) || 
       !providers.contains_key(registry.default_provider.get(group_name).unwrap()) {
        if let Some(first_key) = providers.keys().min() {
            registry.default_provider.insert(group_name.to_string(), first_key.clone());
        }
    }

    (
        providers,
        SchemaValidatorDiscoveryResult {
            group: group_name.to_string(),
            loaded: loaded_repr,
            errors: HashMap::new(),
        },
    )
}

/// Manually register a schema validator provider (useful for tests / custom wiring).
pub fn set_schema_validator(
    validator: Arc<dyn ISchemaValidator + Send + Sync>,
    name: Option<&str>,
    group: Option<&str>,
    make_default: bool,
) {
    let name = name.unwrap_or("manual");
    let group_name = group.unwrap_or(DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP);
    
    let mut registry = get_registry().lock().unwrap();
    registry.attempted_groups.insert(group_name.to_string());
    registry.providers
        .entry(group_name.to_string())
        .or_insert_with(HashMap::new)
        .insert(name.to_string(), validator);
    
    if make_default {
        registry.default_provider.insert(group_name.to_string(), name.to_string());
    }
}

/// Get a schema validator provider.
pub fn get_schema_validator(
    name: Option<&str>,
    group: Option<&str>,
) -> Option<Arc<dyn ISchemaValidator + Send + Sync>> {
    let group_name = group.unwrap_or(DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP);
    let (providers, _) = discover_schema_validators(Some(group_name), false);
    
    if providers.is_empty() {
        return None;
    }

    if let Some(n) = name {
        return providers.get(n).cloned();
    }

    // Get default provider
    let registry = get_registry().lock().unwrap();
    if let Some(default_name) = registry.default_provider.get(group_name) {
        if let Some(provider) = providers.get(default_name) {
            return Some(provider.clone());
        }
    }

    // Fall back to deterministic choice
    if let Some(first_key) = providers.keys().min() {
        return providers.get(first_key).cloned();
    }

    None
}

/// List discovered provider names.
pub fn available_schema_validators(group: Option<&str>) -> Vec<String> {
    let group_name = group.unwrap_or(DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP);
    let (providers, _) = discover_schema_validators(Some(group_name), false);
    let mut names: Vec<String> = providers.keys().cloned().collect();
    names.sort();
    names
}

File: src/exonware/xwsystem/validation/schema_discovery.py
#!/usr/bin/env python3
#exonware/xwsystem/src/exonware/xwsystem/validation/schema_discovery.py
"""
Schema validator discovery (plugin entry points).
This module provides a lightweight "plugin" mechanism (via Python packaging
entry points) to discover an implementation of
`xwsystem.validation.contracts.ISchemaProvider`.
Why this exists:
- Avoids hard dependency cycles (e.g., xwdata <-> xwschema)
- Keeps xwsystem as the stable contract owner
- Enables optional providers to be installed and discovered at runtime
Entry point group:
  - xwsystem.schema_validators
Providers should register an entry point that resolves to either:
- an ISchemaProvider instance
- a class (constructible with no args) returning an ISchemaProvider instance
- a callable (factory) returning an ISchemaProvider instance
"""

from __future__ import annotations
from dataclasses import dataclass
import threading
from typing import Any, Optional
from ..config.logging_setup import get_logger
from .contracts import ISchemaProvider
logger = get_logger("xwsystem.validation.schema_discovery")
DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP = "xwsystem.schema_validators"
@dataclass(frozen=True)


class SchemaValidatorDiscoveryResult:
    """Structured discovery information for observability/debugging."""
    group: str
    loaded: dict[str, str]  # name -> provider repr
    errors: dict[str, str]  # name -> error string
_lock = threading.RLock()
_attempted_groups: set[str] = set()
_providers_by_group: dict[str, dict[str, ISchemaProvider]] = {}
_default_provider_by_group: dict[str, str] = {}


def _as_schema_provider(obj: Any) -> Optional[ISchemaProvider]:
    """Coerce an entry-point loaded object into an ISchemaProvider instance."""
    if obj is None:
        return None
    # 1) Direct instance
    try:
        if isinstance(obj, ISchemaProvider):
            return obj
    except Exception:
        # Protocol runtime checks can raise if obj is weird; treat as not valid.
        return None
    # 2) Class -> instantiate
    if isinstance(obj, type):
        try:
            inst = obj()
        except Exception:
            return None
        try:
            return inst if isinstance(inst, ISchemaProvider) else None
        except Exception:
            return None
    # 3) Factory callable -> call
    if callable(obj):
        try:
            inst = obj()
        except Exception:
            return None
        try:
            return inst if isinstance(inst, ISchemaProvider) else None
        except Exception:
            return None
    return None


def discover_schema_validators(
    *,
    group: str = DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP,
    force: bool = False,
) -> tuple[dict[str, ISchemaProvider], SchemaValidatorDiscoveryResult]:
    """
    Discover schema validator providers from entry points.
    Args:
        group: Entry-point group to search.
        force: If True, re-run discovery even if already attempted.
    Returns:
        (providers, discovery_result)
    """
    with _lock:
        already_attempted = group in _attempted_groups
        if already_attempted and not force:
            providers = dict(_providers_by_group.get(group, {}))
            loaded_repr = {k: repr(v) for k, v in providers.items()}
            return providers, SchemaValidatorDiscoveryResult(group=group, loaded=loaded_repr, errors={})
    loaded: dict[str, ISchemaProvider] = {}
    loaded_repr: dict[str, str] = {}
    errors: dict[str, str] = {}
    try:
        from importlib.metadata import entry_points
        eps = entry_points(group=group)
    except Exception as e:
        # Cache attempted to avoid repeated expensive failures.
        with _lock:
            _attempted_groups.add(group)
            _providers_by_group.setdefault(group, {})
        logger.debug(f"Schema validator discovery failed for group '{group}': {e}")
        return {}, SchemaValidatorDiscoveryResult(group=group, loaded={}, errors={"__discovery__": str(e)})
    for ep in eps:
        name = getattr(ep, "name", "<unknown>")
        try:
            obj = ep.load()
            inst = _as_schema_provider(obj)
            if inst is None:
                errors[name] = "Entry point did not resolve to an ISchemaProvider (instance/class/factory)."
                continue
            loaded[name] = inst
            loaded_repr[name] = repr(inst)
        except Exception as e:
            errors[name] = str(e)
    with _lock:
        _attempted_groups.add(group)
        _providers_by_group[group] = dict(loaded)
        # Preserve existing default selection if still valid; else choose first.
        if group not in _default_provider_by_group or _default_provider_by_group[group] not in loaded:
            if loaded:
                _default_provider_by_group[group] = sorted(loaded.keys())[0]
    if errors:
        logger.debug(f"Schema validator discovery had errors for group '{group}': {errors}")
    return dict(loaded), SchemaValidatorDiscoveryResult(group=group, loaded=loaded_repr, errors=errors)


def set_schema_validator(
    validator: ISchemaProvider,
    *,
    name: str = "manual",
    group: str = DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP,
    make_default: bool = True,
) -> None:
    """Manually register a schema provider (useful for tests / custom wiring)."""
    with _lock:
        _attempted_groups.add(group)
        _providers_by_group.setdefault(group, {})
        _providers_by_group[group][name] = validator
        if make_default:
            _default_provider_by_group[group] = name


def get_schema_validator(
    *,
    name: Optional[str] = None,
    group: str = DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP,
) -> Optional[ISchemaProvider]:
    """
    Get a schema provider (validator).
    Args:
        name: Optional provider name; if omitted, returns the group's default provider.
        group: Entry-point group.
    Returns:
        ISchemaProvider instance or None if no providers available.
    """
    providers, _ = discover_schema_validators(group=group, force=False)
    if not providers:
        return None
    if name:
        return providers.get(name)
    with _lock:
        default_name = _default_provider_by_group.get(group)
    if default_name and default_name in providers:
        return providers[default_name]
    # Fall back to deterministic choice
    return providers[sorted(providers.keys())[0]]


def available_schema_validators(
    *,
    group: str = DEFAULT_SCHEMA_VALIDATOR_ENTRYPOINT_GROUP,
) -> list[str]:
    """List discovered provider names."""
    providers, _ = discover_schema_validators(group=group, force=False)
    return sorted(providers.keys())
